### ISSUE TITLE
Migration to OrcJIT

### ISSUE THREAD

**Issue by sklam** [numba/numba#2290]
In llvm-dev ML: https://groups.google.com/forum/#!searchin/llvm-dev/engine$20removeModule$20leak%7Csort:relevance/llvm-dev/h8PNYl4uXVE/esP6HFHVDgAJ

* MCJIT removeModule is probably leaking; ORC version is good
* MCJIT will be deprecated

But, OrcMCJITReplacement is not good enough.  See discussion in https://github.com/numba/llvmlite/pull/245

**Comment by seibert** [comment]
#570 is related to the memory leak.

**Comment by stuartarchibald** [comment]
xref: https://github.com/numba/numba/issues/4026#issuecomment-768952398 Quoting @lhames :

> @esc Julia, Postgres and Clasp are the larger clients that I've worked with on upgrades. There is some documentation on moving from OrcV1 to OrcV2 [1], but it's not particularly helpful in this case as it mostly deals with the differences between those two APIs.
> 
> The short version of the MCJIT -> OrcV2 upgrade story is: LLJIT is the new MCJIT. The APIs are different, but the all the concepts from MCJIT have counterparts in LLJIT. You can find several LLJIT usage examples in llvm/examples/OrcV2Examples directory, including a few showing how to use the new C APIs.
> 
> I took a quick look at numba and llvmlite -- I think you would need new llvmlite bindings for the APIs in llvm/include/llvm-c/Orc.h, llvm/include/llvm-c/LLJIT.h and llvm/include/llvm-c/OrcEE.h. In numba/numba/core/codegen.py you would just need to use an LLJIT instance to back your JitEngine class, rather than an ExecutionEngine.
> 
> That said, my other recommendation after looking at your code would be to wait until the LLVM 13 release at the end of this year: The APIs that you have been living on are super stable at this point, whereas LLJIT still has some churn from version to version. I foresee more during LLVM 13 development as we roll out the new ORC runtime. Without a PPC JITLink implementation to enable the new features an upgrade now might cause pain for limited gain*.
> 
> [1] https://github.com/llvm/llvm-project/blob/main/llvm/docs/ORCv2.rst#transitioning-from-orcv1-to-orcv2
> 
>     OrcV2 does support concurrent compilation of JIT'd code, lazy compilation, and has better error checking, even when using RuntimeDyld. Whether those alone are useful enough to you to justify the upgrade cost I don't know.
>

**Comment by stuartarchibald** [comment]
High level things which would be nice to get out of such an upgrade include:

* Threadsafe LLVM APIs (if this is possible), this would help reduce the need for Numba's global compiler lock and llvmlite's LLVM API lock.
* Concurrent and lazy compilation opportunities, particularly the concurrent part.
* Support for `perf(1)` (this is really important for https://github.com/numba/numba/issues/5028), I've hacked it into `MCJIT` for R&D purposes in that ticket.

**Comment by lhames** [comment]
> Threadsafe LLVM APIs (if this is possible), this would help reduce the need for Numba's global compiler lock and llvmlite's LLVM API lock.

OrcV2's addModule and lookup APIs are threadsafe. The C API inherits this thread safety.

> Concurrent and lazy compilation opportunities, particularly the concurrent part.

Configuration for concurrent compiles has not been plumbed up to the C API yet -- nobody has asked for it yet and I'm focused on runtime and JIT development. Please let me know when you start looking at this though and I'll be happy to work with you on the C API design.

> Support for perf(1) (this is really important for #5028), I've hacked it into MCJIT for R&D purposes in that ticket.

I believe it should be possible to add a perf JITEventListener to RuntimeDyld via the existing OrcEE APIs.

Side note: We have not written a JITLink Plugin version of the Perf event listener, but I doubt it would be prohibitively difficult. This will only become relevant if/when you consider switching to a PPC JITLink implementation.

**Comment by lhames** [comment]
Answering an older comment:

> MCJIT removeModule is probably leaking

OrcV2's API for this is ResourceTracker::remove, and it's C API counterpart:
LLVMErrorRef LLVMOrcResourceTrackerRemove(LLVMOrcResourceTrackerRef RT);

I know Postgres are using that C API in their OrcV2 implementation -- we ironed out a couple of leaks during bring-up of the API but it seems to be working well now.

**Comment by abebeos** [comment]
@stuartarchibald , maybe update the issue-text/title (or open a new one), as it describes outdated stuff (see: https://github.com/numba/llvmlite/pull/245#issuecomment-771207184)

**Comment by esc** [comment]
@abebeos thank you for pointing this out, I will take the OrcV2 issue (not any specific GitHub issue, but the issue of how to move forward with OrcV2) into the developer meeting next week. Hopefully, we can update any outdated issues subsequently.

**Comment by esc** [comment]
@abebeos we discussed this in the issue triage today and we came to the conclusion that it will be some time (months?) before we can think about tackling a migration away from MCJIT. I am not sure what to change the title of this issue however, maybe "move away from MCJIT"?

**Comment by abebeos** [comment]
Possibly just:

## Migrate from MCJIT to OrcV2

(but you would need to  update the issue-text, too, as it is outdated/obsolete. Alternatively just close and create a new issue)

**Comment by esc** [comment]
Sure, why not, new issue is at: https://github.com/numba/numba/issues/6707
